File: backend/issue-service/src/main/java/com/example/issueservice/services/AuthService.java
package com.example.issueservice.services;

import com.example.issueservice.config.PermissionCacheReader;
import com.example.issueservice.dto.models.enums.ActionType;
import com.example.issueservice.dto.models.enums.EntityType;
import com.example.issueservice.dto.response.UserPermissionsResponse;
import com.example.issueservice.exception.AccessDeniedException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {
    private final PermissionCacheReader cacheReader;

    public void hasPermission(Long userId, Long projectId, EntityType entity, ActionType action) {
        UserPermissionsResponse perms = cacheReader.getUserPermissions(userId, projectId);
        if(perms.permissions().contains(entity.name() + ":" + action.name())) {
            throw new AccessDeniedException("User has no permission to view issues");
        }
    }

    public UserPermissionsResponse getUserPermissions(Long userId, Long projectId) {
        return cacheReader.getUserPermissions(userId, projectId);
    }
}

File: backend/dashboard-service/Backend.Dashboards.Api/Services/AuthService.cs
using Backend.Dashboard.Api.Cache;
using Backend.Shared.DTOs;

namespace Backend.Dashboard.Api.Services
{
    public class AuthService
    {
        private readonly PermissionCacheReader _cacheReader;
        private readonly ILogger<AuthService> _logger;

        public AuthService(PermissionCacheReader cacheReader, ILogger<AuthService> logger)
        {
            _cacheReader = cacheReader;
            _logger = logger;
        }

        public async Task HasPermissionAsync(long userId, long projectId, EntityType entity, ActionType action)
        {
            var perms = await _cacheReader.GetUserPermissionsAsync(userId, projectId);
            var requiredPermission = $"{entity}:{action}";

            if (!perms.Permissions.Contains(requiredPermission))
            {
                throw new UnauthorizedAccessException(
                $"User {userId} has no {action} permission on {entity} in project {projectId}");
            }
        }

        public Task<UserPermissionsResponse> GetUserPermissionsAsync(long userId, long projectId)
        {
            return _cacheReader.GetUserPermissionsAsync(userId, projectId);
        }

        public void HasPermission(long userId, long projectId, EntityType entity, ActionType action)
        {
            HasPermissionAsync(userId, projectId, entity, action).GetAwaiter().GetResult();
        }
    }
}
